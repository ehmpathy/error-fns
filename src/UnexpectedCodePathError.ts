import { HelpfulError, HelpfulErrorMetadata } from './HelpfulError';

/**
 * UnexpectedCodePathError errors are used to explicitly declare that we've reached a code path that should never have been reached
 */
export class UnexpectedCodePathError extends HelpfulError {
  constructor(message: string, metadata?: HelpfulErrorMetadata) {
    super(['UnexpectedCodePathError: ', message].join(''), metadata);
  }
}
