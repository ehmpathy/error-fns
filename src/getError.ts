/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import { isAFunction, isAPromise } from 'type-fns';

export class NoErrorThrownError extends Error {}

/**
 * a method which returns the error thrown from executing the logic passed into it
 *
 * usecases
 * - testing that a some logic throws an error
 *
 * note
 * - uses type overloading, so if you pass in a promise or async function, you'll be given a promise; otherwise, sync
 */
export function getError<T = Error>(
  from: () => Promise<unknown>,
): Promise<T | NoErrorThrownError>;
export function getError<T = Error>(
  from: Promise<unknown>,
): Promise<T | NoErrorThrownError>;
export function getError<T = Error>(
  from: () => unknown,
): T | NoErrorThrownError;
export function getError<T = Error>(
  from: Promise<unknown> | (() => Promise<unknown>) | (() => unknown),
) {
  // run everything in a try catch, to handle the case where the function throws an error immediately (does not return a promise)
  try {
    // execute the function to get the result, if its a function
    const result = isAFunction(from) ? from() : from;

    // handle the result = promise case
    if (isAPromise(result))
      return result
        .then(() => {
          throw new NoErrorThrownError();
        })
        .catch((error) => error as T | NoErrorThrownError);

    // handle the result = nonpromise case
    throw new NoErrorThrownError(); // should have never reached here, as it should have thrown when we first called it
  } catch (error: unknown) {
    return error as T;
  }
}
