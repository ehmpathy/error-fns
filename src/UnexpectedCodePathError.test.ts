import { UnexpectedCodePathError } from './UnexpectedCodePathError';
import { getError } from './getError';

describe('UnexpectedCodePathError', () => {
  it('should produce a helpful, observable error message', () => {
    const error = new UnexpectedCodePathError('no tires on the vehicle', {
      tires: [],
    });
    expect(error).toMatchSnapshot();
  });
  it('should be throwable in a ternary conveniently and precisely', () => {
    const error = getError(() => {
      // this case should not throw
      const customerOne: { phone: string | null } = {
        phone: 'yes',
      };
      const phoneOne =
        customerOne.phone ??
        UnexpectedCodePathError.throw('phone one not found!');

      // but this case should throw
      const customerTwo: { phone: string | null } = {
        phone: null,
      };
      const phoneTwo =
        customerTwo.phone ??
        UnexpectedCodePathError.throw('phone two not found!');
    });
    expect(error).toBeInstanceOf(UnexpectedCodePathError);
    expect(error.message).toContain('phone two not found!');
  });
});
