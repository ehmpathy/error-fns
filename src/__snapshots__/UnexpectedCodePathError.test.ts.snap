// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnexpectedCodePathError should produce a helpful, observable error message 1`] = `
[Error: UnexpectedCodePathError: no tires on the vehicle

{
  "tires": []
}]
`;
